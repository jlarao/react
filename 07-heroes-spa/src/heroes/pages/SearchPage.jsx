import { useLocation, useNavigate } from 'react-router-dom';
import { useForm } from '../../hook/useForm';
import { HeroCard } from '../components';
import queryString from 'query-string';
import { getHeroesByName } from '../helpers/getHeroesByName';

export const SearchPage = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const query = queryString.parse(location.search);

    const { q = '' } = query;

    const heroes = getHeroesByName(q);

    const showSearch = (q.length === 0);
    const showError = (q.length > 0) && heroes.length === 0;

    const {searchText, onInputChange} = useForm({
        searchText: q
    });

    const onSearchSubmin = (event) => {
        event.preventDefault();
        // if( searchText.length <= 1 ) return;

        navigate(`?q=${ searchText }`);

        // console.log({ searchText });
    }
    return (
        <>
            <h1>SearchPage</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Search</h4>
                    <hr />

                    <form onSubmit={onSearchSubmin}>
                        <input 
                            type="text" 
                            placeholder="Find your hero" 
                            className="form-control"
                            name="searchText"
                            autoComplete="off" 
                            value={searchText}
                            onChange={onInputChange}/>


                        <button 
                        className="btn btn-outline-primary mt-1">
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    {/*
                        (q === '') ?
                        <div className="alert alert-primary">
                            Search a hero
                        </div> 
                    : 
                    (heroes.length === 0) &&                     
                        <div className="alert alert-danger">
                            Hero not found <b> {q} </b>
                        </div>

                    */}
                    <div className="alert alert-primary animate__animated animate__fadeIn" style={{ display: (showSearch) ? '' : 'none' }}>
                        Search a hero
                    </div>

                    <div className="alert alert-danger animate__animated animate__fadeIn" style={{ display: (showError) ? '' : 'none' }} aria-label='alert-danger'>
                            Hero not found <b> {q} </b>
                        </div>

                </div>


                { heroes.map( hero => (<HeroCard key={hero.id} {...hero} />)) }

            </div>

        </>
    )
}