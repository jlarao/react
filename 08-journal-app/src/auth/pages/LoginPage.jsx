import { Google } from "@mui/icons-material";
import { Alert, Button, Grid, Link, TextField, Typography } from "@mui/material"
import { Link  as RouterLink} from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import {  useMemo } from "react";
import { startGoogleSignIn, startLoginWithEmailPassword} from "../../store/auth/thunks";

import { AuthLayout } from "../layout/AuthLayout";
import { useForm } from "../../hooks";

const formData = {
    email: '',
    password: '',
}
export const LoginPage = () => {

    const dispatch = useDispatch();
    const { status, errorMessage } = useSelector( state => state.auth);

    const { email, password, onInputChange} = useForm(formData);

    const isAuthenticating = useMemo( () => status === 'cheking', [ status] );

    const onSubmit = (event) => {
        // console.log({ email, password });
        event.preventDefault();
        // dispatch(checkingAuthentication(email, password));
        dispatch( startLoginWithEmailPassword({email, password}));

    }

    const onGoogleSignIn = () =>{
        console.log('onGoogleSignIn');
        dispatch( startGoogleSignIn() );
    }

    // useEffect(() => {
    //     dispatch(checkingAuthentication(email, password));
    // })
    // const errorMessage = 'Email or Password are incorrect';
    
    return (
        
            <AuthLayout title="Login!!!">

                <form onSubmit={ onSubmit } className="animate__animated animate__fadeIn animate__faster">
                    <Grid container>

                        <Grid item xs={ 12 } sx={{ mt: 2 }}
                            display = { !!errorMessage? '' : 'none'}>
                            <Alert severity="error">
                                { errorMessage }
                            </Alert>
                        </Grid>

                        <Grid item xs={ 12 } sx={{ mt: 2 }}>
                            <TextField
                                label="Email"
                                type="email"
                                placeholder="Email"
                                fullWidth
                                name='email'
                                value={ email }
                                onInput={ onInputChange }
                            />
                        </Grid>

                        <Grid item xs={ 12 } sx={{ mt: 2 }}>
                            <TextField
                                label="Password"
                                type="password"
                                placeholder="Password"
                                fullWidth
                                name='password'
                                value={ password }
                                onInput={ onInputChange }
                            />
                        </Grid>

                        <Grid 
                            container spacing={ 2 } 
                            sx={{ mb: 2, mt: 1 }}>
                            
                            {/* <Grid item 
                            xs={ 12 } 
                            display={ !!errorMessage ? '' : 'none' }>
                                <Alert severity="error">
                                    { errorMessage }
                                </Alert>
                            </Grid> */}

                            <Grid item xs={ 12 } sm={ 6 }>
                                <Button 
                                    disabled= { isAuthenticating}
                                    type="submit"
                                    variant="contained" 
                                    fullWidth>
                                    Login
                                </Button>
                            </Grid>

                            
                            <Grid item xs={ 12 } sm={ 6 }>
                                <Button 
                                    disabled = { isAuthenticating}
                                    variant="contained" 
                                    fullWidth
                                    onClick={ onGoogleSignIn}>
                                        <Typography sx={{ mr: 1 }}>
                                            Google
                                        </Typography>                                    
                                </Button>
                            </Grid>
                        </Grid>
                        
                        <Grid container 
                            direction="row" 
                            justifyContent="end">
                            <Link
                                component={ RouterLink } 
                                color="inherit" 
                                to="/auth/register">
                                Create new account
                            </Link>
                        </Grid>
                    </Grid>
                </form>

            </AuthLayout>
        
    )
}