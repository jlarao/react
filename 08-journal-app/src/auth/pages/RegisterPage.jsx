import { Alert, Button, Grid, Link, TextField, Typography } from "@mui/material"
import { AuthLayout } from "../layout/AuthLayout"
import { Link as RouterLink } from "react-router-dom"
import { useForm } from "../../hooks"
import { useMemo, useState } from "react"
import { useDispatch, useSelector } from "react-redux"
import { startCreatingUserWithEmailPassword } from "../../store/auth/thunks"


const formData = {
    email: 'pisujo@gmail.com',
    password: '12345678',
    displayName: 'Pisujo',
}

const formValidations = {
    email : [(value)=> value.includes('@'), 'el correo debe tener una @'],
    password : [(value) => value.length >= 6, 'el password debe de tener mas de 6 letras'],
    displayName : [(value) => value.length >= 1, 'el nombre es obligatorio'],
}
export const RegisterPage = () => {

    const dispatch = useDispatch();

    const[ formSubmitted, setFormSubmitted] = useState(false);
    const {status, errorMessage} =  useSelector( state => state.auth);
    const isCheckingAuthentication = useMemo( () => status === 'checking', [status] );

    const { 
        displayName, email, password, onInputChange, formState,
        isFormValid, displayNameValid, emailValid, passwordValid} = useForm( formData, formValidations );

    const onSubmit = (event) =>{

        event.preventDefault();
        // console.log(formData);
        setFormSubmitted(true);
        
        if(!isFormValid) return;

        dispatch(startCreatingUserWithEmailPassword(formState))
    }
    return (
        
        <AuthLayout title="Crear Cuenta">
            <h1> Form  { isFormValid ? 'Valido' : 'Incorrecto' }</h1>

            <form onSubmit={ onSubmit} className="animate__animated animate__fadeIn animate__faster">
                <Grid container>
                    <Grid item xs={ 12 } sx={{ mt: 2 }}>
                            <TextField
                                label="Nombre Completo"
                                type="text"
                                placeholder="Nombre"
                                fullWidth
                                name = "displayName"
                                value = { displayName}
                                onChange={ onInputChange}
                                error={ !!displayNameValid && formSubmitted}
                                helperText={ displayNameValid}
                            />
                    </Grid>

                    <Grid item xs={ 12 } sx={{ mt: 2 }}>
                        <TextField
                            label="Email"
                            type="email"
                            placeholder="Email"
                            fullWidth
                            name = "email"
                            value = { email}
                            onChange={ onInputChange}
                            error={ !!emailValid && formSubmitted}
                            helperText={ emailValid}
                        />
                    </Grid>

                    <Grid item xs={ 12 } sx={{ mt: 2 }}>
                        <TextField
                            label="Password"
                            type="password"
                            placeholder="Password"
                            fullWidth
                            name = "password"
                            value = { password}
                            onChange={ onInputChange}
                            error={ !!passwordValid && formSubmitted}
                            helperText={ passwordValid}
                        />
                    </Grid>

                    <Grid 
                        container 
                        sx={{ mb: 2, mt: 1 }}>
                        

                        <Grid item 
                            xs={ 12 }
                            display={ !!errorMessage? '' : 'none'}>
                            <Alert severity="error">{ errorMessage}</Alert>
                        </Grid>

                        <Grid item xs={ 12 } >
                            <Button
                                disabled = { isCheckingAuthentication}
                                type="submit"
                                variant="contained" 
                                fullWidth>
                                Crear cuenta
                            </Button>
                        </Grid>
                    </Grid>
                    
                    <Grid container 
                        direction="row" 
                        justifyContent="end">
                            <Typography sx={{ mr: 1 }}> Ya tenes una cuenta?</Typography>

                        <Link
                            component={ RouterLink } 
                            color="inherit" 
                            to="/auth/login">
                            Ingresar
                        </Link>
                    </Grid>
                </Grid>
            </form>

        </AuthLayout>
    
)
}