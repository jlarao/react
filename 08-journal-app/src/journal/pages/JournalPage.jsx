import { AddOutlined, MailOutline } from "@mui/icons-material"
import { IconButton, Typography } from "@mui/material"
import { JournalLayout } from "../Layout/JournalLayout"
import { NoteView, NothingSelectedView } from "../views"

export const JournalPage = () => {

    return (
        <JournalLayout>
            {/* <Typography variant="p">
                Lorem ipsum dolor, sit amet consectetur adipisicing elit. Expedita velit ab delectus pariatur ex nesciunt neque odit quasi quam dignissimos, omnis tempore sint, distinctio magni, laudantium doloremque voluptates eos enim.
            </Typography> */}
            <NothingSelectedView/>
            {/* <NoteView/> */}
            <IconButton
                size="large"
                sx = {{ 
                    color: 'white',
                    backgroundColor: 'error.main',
                    ':hover': { backgroundColor: 'error.main', opacity: 0.9 },
                    position: 'fixed',
                    right: 50,
                    bottom: 50
                }} 
                >
                
                <AddOutlined sx={{ fontSize: 30 }} />

            </IconButton>
        </JournalLayout>
    )
}