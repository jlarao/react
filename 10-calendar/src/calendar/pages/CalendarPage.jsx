import { CalendarEvent, CalendarModal, FabAddNew, FabDelete, Navbar } from "../";

import { Calendar } from 'react-big-calendar'
import { localizer, getMessagesES } from '../../helper';

import 'react-big-calendar/lib/css/react-big-calendar.css';
import { useState } from "react";
import { useCalendarStore, useUiStore } from "../../hooks";

export const CalendarPage = () => {
    const { events, setActiveEvent } = useCalendarStore();
    const {openDateModal } = useUiStore();
    const [lastView, setLastView] = useState(localStorage.getItem('lastView') || 'week');
    const eventStyleGetter = (event, start, end, isSelected) => {

        const style = {
            backgroundColor: '#347CF7',
            borderRadius: '0px',
            opacity: 0.8,
            color: 'white',
            border: '0px',
            display: 'block'
        }
        return { style }
    }

    const onDoubleClick = (e) => {
        console.log(e);
        openDateModal();
    }

    const onSelect = (e) => {
        console.log(e);
        setActiveEvent(e);
    }

    const onViewChange = (e) => {
        console.log(e);
        localStorage.setItem('lastView', e);
        // setLastView(e);
    }

    return (
        <>
            <Navbar />
            <h1>Calendar Page 1</h1>

            <Calendar
                culture="es"
                localizer={localizer}
                events={events}
                defaultView={lastView}
                startAccessor="start"
                endAccessor="end"
                style={{ height: 500 }}
                messages={getMessagesES()}
                eventPropGetter={eventStyleGetter}
                components={{
                    event: CalendarEvent
                }}
                onDoubleClickEvent={ onDoubleClick }
                onSelectEvent={ onSelect }
                onView={ onViewChange }
                />

                <CalendarModal />
                <FabAddNew />
                <FabDelete />
        </>
    )
}