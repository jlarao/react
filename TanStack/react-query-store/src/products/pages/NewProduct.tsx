import { Button, Image, Input, Textarea } from "@nextui-org/react";
import { Controller, SubmitHandler, useForm } from "react-hook-form";
import { useProductMutation } from "../hooks/useProductMutation";

interface FormInputs {
    title:       string;
    price:       number;
    description: string;
    category:    string;
    image:       string;
}
export const NewProduct = () => {
  const productMutation = useProductMutation();

  const { control ,  handleSubmit, watch } = useForm<FormInputs>({
    defaultValues: {
      title:       "Teclado",
      price:       150,
      description: 'EL-993384'+
      ''+
      'Teclado con 8 teclas multimedia. Cuenta con sistema de membrana para un tecleo silencioso y conector USB 2.0.'+
      ''+
      ''+
      'ESPECIFICACIONES TÉCNICAS'+
      ''+
      '• Conector: USB'+
      '• Longitud de cable: 1.40 m'+
      '• Mecanismo de teclas: Membrana'+
      '• Alimentación: 5 Vcc 70 mA'+
      '• Compatibilidad: Windows® XP /Vista / 7 & Mac OS®',
      category:    "men's clothing",
      image:       "https://easyline.me/cdn/shop/products/EL-993384_01C.png?v=1656340663&width=1445",
    }
  });

  const newImage = watch('image');
  
  const onSubmit:SubmitHandler<FormInputs> = (data) => {
    productMutation.mutate(data);
    
  };

  return (
    <div className="w-full flex-col">
      <h1 className="text-2xl font-bold">Nuevo producto</h1>

      <form onSubmit={handleSubmit(onSubmit)} className="w-full">

        <div className="flex justify-around items-center">
          
          <div className="flex-col w-[500px]">
            <Controller 
              control={control}
              name="title"
              rules={{ required: true }}
              render= { ({field}) =>(
                <Input value={field.value} onChange={field.onChange} className="mt-2" type="text" label="Titulo del producto" />
              )}
            />

            <Controller 
              control={control}
              name="price"
              rules={{ required: true }}
              render= { ({field}) =>(
                <Input value={field.value?.toString()} onChange={ev => field.onChange( +ev.target.value)} className="mt-2" type="number" label="Precio del producto" />
            )}
            />

            <Controller 
              control={control}
              name="image"
              rules={{ required: true }}
              render= { ({field}) =>(
                <Input value={field.value} onChange={field.onChange} className="mt-2" type="url" label="Url del producto" /> 
              )}
            />

            <Controller 
              control={control}
              name="description"
              rules={{ required: true }}
              render= { ({field}) =>(
                <Textarea value={field.value} onChange={field.onChange} className="mt-2" label="Descripcion del producto" />
              )}
            />

            <Controller 
              control={control}
              name="category"
              rules={{ required: true }}
              render= { ({field}) =>(
                <select value={field.value} onChange={field.onChange} className="rounded-md p-3 mt-2 bg-gray-800 w-full">
                <option value="men's clothing">Men's clothing</option>
                <option value="women's clothing">Women's clothing</option>
                <option value="jewelery">Jewelery</option>
                <option value="electronics">Electronics</option>
              </select>
              )}
            />
            

            <br />
            <Button 
              type="submit"
              isDisabled={productMutation.isPending}
              className="mt-2" color="primary">
                {productMutation.isPending }
                Crear</Button>
          </div>

          <div className="bg-white rounded-2xl p-10 flex items-center" style={{
            width: '500px',
            height: '600px',
          }}>

            <Image
              src={newImage }
            />
          </div>
          
        </div>


      </form>

    </div>
  )
}