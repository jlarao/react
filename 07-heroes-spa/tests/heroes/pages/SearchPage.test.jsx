import queryString from 'query-string';
import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockUseNavigate
}))
describe('test search page', () => {
    
    beforeAll(() => jest.clearAllMocks());

    test('debe de mostrarse correctamente con valores por defecto', () => {
        const { container } = render(
        <MemoryRouter>
            <SearchPage />
        </MemoryRouter>
       ); 


       expect( container ).toMatchSnapshot();
       
    //    screen.debug();
    });

    test('debe de mostrar a batman con query', () => {
        const { container } = render(
        <MemoryRouter initialEntries={['/search?q=batman']}>
            <SearchPage />
        </MemoryRouter>
       ); 


        const inputValue = screen.getByRole('textbox');
        expect( inputValue.value ).toBe('batman');

        const image = screen.getByRole('img');
        expect(image.src).toContain('/assets/heroes/dc-batman.jpg');


        const div = screen.getByLabelText('alert-danger');
        expect(div.style.display).toBe('none');
    //    screen.debug();
    //    console.log(div.style);
    });

    test('debe de mostrar un error si no se encuentra el hero', () => {
        render(
            <MemoryRouter initialEntries={['search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        );
        const div = screen.getByLabelText('alert-danger');
        expect(div.style.display).toBe('');

        // screen.debug();
    });

    test('debe de llamar el push del navigate', () => {

            const inputValue = 'superman';
             render(
                <MemoryRouter  initialEntries={['/search']}>
                    <SearchPage />
                </MemoryRouter>
             );

             const input = screen.getByRole('textbox');
             fireEvent.change(
                input , {
                    target : 
                    {name : 'searchText' , value : inputValue}});

            const buton  = screen.getByRole('button');
            fireEvent.click(buton );
            //  screen.debug();

             expect( mockUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
    });

});