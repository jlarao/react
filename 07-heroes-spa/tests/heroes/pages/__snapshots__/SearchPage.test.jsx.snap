// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test search page debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    SearchPage
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        Hero not found 
        <b>
           
           
        </b>
      </div>
    </div>
  </div>
</div>
`;
